src/ctap2/commands/mod.rs

```diff
@@ -97,7 +97,8 @@ pub(crate) trait PinAuthCommand {
     fn client_data_hash(&self) -> ClientDataHash;
     fn unset_uv_option(&mut self);
     fn determine_pin_auth<D: FidoDevice>(&mut self, dev: &mut D) -> Result<(), AuthenticatorError> {
-        if !dev.supports_ctap2() {
+        // Authenticator info should have been set if the device supports ctap2
+        if !dev.supports_ctap2() || dev.get_authenticator_info().unwrap().options.client_pin.is_none() {
```

Comment:
Let's avoid `unwrap` here. There's some logic for late initialization of the AuthenticatorInfo in `establish_shared_secret` which makes me think there may be situations where it fails to initialize.

Reply:
Interesting, I indeed missed that it does that. On that note, I also missed that there is already logic to handle devices that don't support client_pin in `calculate_pin_auth` but that logic is never reached in my case as `establish_shared_secret` already requires support for `client_pin` (constructing `GetKeyAgreement` checks the supported pin protocols). It is also worth noting that the logic in `establish_shared_secret` for obtaining the authenticator info is "just to be safe" and I don't think it really makes sense to call these functions without first calling `init` (as is currently done in the statemachine for example).  I suggest the following changes:

- Don't return the authenticator info from `establish_shared_secret` and don't attempt to obtain it in that function if it is not yet set. I would consider calling this function without calling `init` to be a programmer mistake (either handled by using `unwrap`/`expect` or by returning and error to the caller). Current uses all go through `StateMachineCtap2` which calls `init` on device selection: https://github.com/mozilla/authenticator-rs/blob/98b120bd66ace4b88a0314bbd8ab09bd3df7d30c/src/statemachine.rs#L322
- Obtain the authenticator info separately in `calculate_pin_auth` (handled in the same way as in the case of  `establish_shared_secret`) and only after checking that `client_pin` is supported should it call `establish_shared_secret`. Other call sites of `establish_shared_secret` (there's two in statemachine.rs) should be updated in a similar way

This could be fixed in a cleaner way by making CTAP1 and CTAP2 different traits rather than combining them in the `FidoDevice` trait for example, with the CTAP2 trait ensuring the authenticator info was collected. But I consider that out of scope for this PR.

Reply:
I'm currently looking into other UV-methods like fingerprint-sensors as part of CTAP2.1 and we'll very likely need to change quite a few things in this area anyways (also for NFC-support as to the order of when collecting the PIN).
So my suggestion would be to keep the changes in this PR as minimal as possible (since we'll probably have to change it soon anyways).
So maybe just something like using `map_or_else()` instead of `unwrap()`?

Reply:
Oh and btw.: Could you maybe send me the full AuthenticatorInfo that is returned by mooltipass?
That would be very helpful to ensure we are doing the right thing here.

You can get this by running
```
RUST_LOG=authenticator::transport=debug cargo run --example ctap2
```
In there should be a line like:
```
[2023-03-02T10:19:15Z DEBUG authenticator::transport] "/dev/hidraw6" infos: AuthenticatorInfo { versions: ["U2F_V2", "FIDO_2_0", "FIDO_2_1_PRE", "FIDO_2_1"], .... }
```

Reply:
```
AuthenticatorInfo {
    versions: ["FIDO_2_0"],
    extensions: [],
    aaguid: AAGuid(6db042d0-61af-404c-a887-e72e09ba7eb4),
    options: AuthenticatorOptions {
        platform_device: false,
        resident_key: true,
        client_pin: None,
        user_presence: true,
        user_verification: Some(true)
    },
    max_msg_size: Some(1024),
    pin_protocols: [],
    max_credential_count_in_list: None,
    max_credential_id_length: None,
    transports: None,
    algorithms: None,
    max_ser_large_blob_array: None,
    force_pin_change: None,
    min_pin_length: None,
    firmware_version: None,
    max_cred_blob_length: None,
    max_rpids_for_set_min_pin_length: None,
    preferred_platform_uv_attempts: None,
    uv_modality: None, certifications: None,
    remaining_discoverable_credentials: None,
    vendor_prototype_config_commands: None
}
```